### PR TITLE
[FLINK-32380] Support Java Records with PojoTypeInfo/Serializer

### PR DESCRIPTION
**Pull request by gyfora**
## What is the purpose of the change

Extend the current TypeExtraction + PojoSerializer logic to support handling Java Record classes.
The main difference is the extra codepath needed to instantiate records through the full constructor instead of using field setters to assemble it.

### Assumptions / limitations

The record logic assumes that the record objects can be instantiated correctly with the canonical constructor. 
Which means that the following record will not be serialised correctly, but I think that's okay because it's clearly not the intended way of using java records.

```
public record MyRecord(int a, int b) {
        public MyRecord(int a, int b) {
            this.b = a;
            this.a = b;
        }
    }
```

### Benchmarks

Performance was measured with the flink benchmark utility to avoid any potential regressions due to the new code in the serializer:

```
# 1.17.1

PojoSerializationBenchmark.copyPojo       thrpt   30  4523,595 ± 14,500  ops/ms
PojoSerializationBenchmark.copyPojoReuse  thrpt   30  4508,138 ± 11,756  ops/ms
PojoSerializationBenchmark.copyView       thrpt   30   533,934 ±  2,142  ops/ms
PojoSerializationBenchmark.readPojo       thrpt   30   902,677 ±  7,109  ops/ms
PojoSerializationBenchmark.writePojo      thrpt   30   671,874 ± 18,102  ops/ms

SerializationFrameworkAllBenchmarks.serializerHeavyString              thrpt   30   193,195 ±   0,696  ops/ms
SerializationFrameworkAllBenchmarks.serializerPojo                     thrpt   30  1166,849 ± 323,670  ops/ms
SerializationFrameworkAllBenchmarks.serializerPojoWithoutRegistration  thrpt   30   822,910 ±  24,877  ops/ms

# 1.18-SNAPSHOT

PojoSerializationBenchmark.copyPojo       thrpt   30  4626,837 ± 14,787  ops/ms
PojoSerializationBenchmark.copyPojoReuse  thrpt   30  4604,001 ±  6,462  ops/ms
PojoSerializationBenchmark.copyView       thrpt   30   509,084 ±  6,469  ops/ms
PojoSerializationBenchmark.readPojo       thrpt   30   851,225 ±  6,859  ops/ms
PojoSerializationBenchmark.writePojo      thrpt   30   688,037 ±  7,798  ops/ms

SerializationFrameworkAllBenchmarks.serializerHeavyString              thrpt   30   193,136 ±   0,649  ops/ms
SerializationFrameworkAllBenchmarks.serializerPojo                     thrpt   30  1500,717 ± 315,250  ops/ms
SerializationFrameworkAllBenchmarks.serializerPojoWithoutRegistration  thrpt   30  1163,223 ± 317,971  ops/ms

# Java Record serializer support

PojoSerializationBenchmark.copyPojo       thrpt   30  4558,056 ± 40,014  ops/ms
PojoSerializationBenchmark.copyPojoReuse  thrpt   30  4521,589 ± 11,021  ops/ms
PojoSerializationBenchmark.copyView       thrpt   30   504,593 ±  6,722  ops/ms
PojoSerializationBenchmark.readPojo       thrpt   30   841,544 ±  6,781  ops/ms
PojoSerializationBenchmark.writePojo      thrpt   30   670,677 ± 13,041  ops/ms

SerializationFrameworkAllBenchmarks.serializerHeavyString              thrpt   30   192,511 ±   0,607  ops/ms
SerializationFrameworkAllBenchmarks.serializerPojo                     thrpt   30  1495,423 ± 316,010  ops/ms
SerializationFrameworkAllBenchmarks.serializerPojoWithoutRegistration  thrpt   30   806,237 ±  40,851  ops/ms
```

Comparing Pojo vs Record serialisation:
```
RecordSerializationBenchmark.copyPojo         thrpt   30  6825.981 ± 32.221  ops/ms
RecordSerializationBenchmark.copyRecord       thrpt   30  5949.899 ± 94.196  ops/ms
RecordSerializationBenchmark.copyPojoReuse    thrpt   30  6757.649 ± 91.903  ops/ms
RecordSerializationBenchmark.copyRecordReuse  thrpt   30  5974.339 ± 51.605  ops/ms
RecordSerializationBenchmark.copyPojoView     thrpt   30   929.945 ±  3.241  ops/ms
RecordSerializationBenchmark.copyRecordView   thrpt   30   916.406 ± 53.823  ops/ms
RecordSerializationBenchmark.readPojo         thrpt   30  1070.375 ±  5.260  ops/ms
RecordSerializationBenchmark.readRecord       thrpt   30  1041.910 ± 11.632  ops/ms
RecordSerializationBenchmark.writePojo        thrpt   30  1225.184 ±  3.824  ops/ms
RecordSerializationBenchmark.writeRecord      thrpt   30  1234.543 ±  4.738  ops/ms
```

### Compatibility

1. Compatible as-is with previous POJO with same fields  (backward and forward) -> Supports converting previous POJOs to Java records
2. Supports same schema evolution as POJOs before (reorder, add/remove fields) etc

## Brief change log

  - *Create PojoTypeInfo for Records in the TypeExtractor*
  - *Add specialised ser/de copy logic to PojoSerializer*
  - *Tests*

## Verifying this change

Existing Pojo tests cover the serializer against regression and we have introduced a new unit test that should only be run on the `java17-target` profile to test the record logic. The pom magic is necessary to not even compile this class in other profiles as `record` declarations would fail on earlier java versions.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not documented

### CODE REVIEW

**Comment by gyfora** [comment]
I just realised that the current implementation won't work with schema evolution as we have to "remap" the parameter indexes for the constructor afterwards, also defining new primitive fields can be problematic.

File: tools/maven/suppressions-core.xml
Comment (XComp) [reply tools/maven/suppressions-core.xml:126]: When looking for the reason for this weird checkstyle rule, I found https://github.com/checkstyle/checkstyle/issues/8023. I verified that this is actual an issue in our current setup.

It's not an issue for this PR because we don't need this feature of records. But I'm wondering whether we should start collecting reasons like that to initiate a discussion on upgrading the checkstyle version. 🤔

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks for creating this PR, @gyfora . I managed to find some time to go over it. Some of the comments appear to be a bit too nit-picky. So, feel free to ignore them if you find them unreasonable.

From a conceptual point of view I feel like we could implement the `JavaRecordSerializationHelper` as a builder factory. I have this proposal in one of the comments below. This would improve the readability of the code. PTAL

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java:72]: That's a very nice idea, it should make this cleaner :)

File: flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/Java17PojoRecordSerializerTest.java
Comment (gyfora) [reply flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/Java17PojoRecordSerializerTest.java:53]: This was unintentional, good catch

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java:37]: Will do!

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java:2079]: makes sense

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:125]: makes sense to make this more explicit, and remove isRecord

**Comment by gyfora** [comment]
Thanks for the detailed comments @XComp , I will try to address them and update the PR next week :)

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java:83]: I would prefer to keep the exception consistent/wrapped

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java:111]: we have the null fields after restoring the deserialiser for anything that was removed, I think we need this.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java:125]: As I noted in the other comment it's not the user passing null it may be passed when restoring after migrating to a pojo without that field

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java:129]: Makes sense, simplifies the logic

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java
Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordSerializationHelper.java:111]: I think that it's worth mentioning it in a comment then

File: pom.xml
Comment (gyfora) [reply pom.xml:1108]: Unfortunately it's not enough to use java 17 but we also need to set the compilation target which is not set in the `java17` profile only in `java17-target`separately

File: pom.xml
Comment (gyfora) [reply pom.xml:210]: I really wanted to avoid this overhead unless the community really hates the prefix based approach :) we already have way too many modules

**Comment by gyfora** [comment]
@XComp I tried to address all of your comments, please take a look when you have some time

**Comment by gyfora** [comment]
A problem with the current PR/test-setup is that we don't actually run the CI using Java 17 (or java17-target profile) this means that the new tests are not executed on azure. This may actually point toward extracting them into a separate module and setting up some new tests for that on azure.

@XComp @zentol what do you think?

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks, @gyfora for addressing my comments. The changes look good. I have a few minor comments, though. PTAL

> A problem with the current PR/test-setup is that we don't actually run the CI using Java 17 (or java17-target profile) this means that the new tests are not executed on azure. This may actually point toward extracting them into a separate module and setting up some new tests for that on azure.
>
> I am actually a bit confused about the azure java version setup and how to set these tests up properly.
>
> @XComp @zentol what do you think?

There's a JDK 17 profile (analogously to JDK 11) that only runs in the nightly builds (see [the most-recent one from master](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=53913&view=results)). The entire Flink test suite will be triggered with this profile. That should also include the tests of this PR. I this where your confusion is coming from or am I missing something here? :thinking: 

Or are you concerned that the tests are not executed in this PR right now and that they are only executed after merging the PR to `master`? That could be temporarily changed in your branch by updating the profile of [tools/azure-pipelines/build-apache-repo.yml:72](https://github.com/apache/flink/blob/9b63099964b36ad9d78649bb6f5b39473e0031bd/tools/azure-pipelines/build-apache-repo.yml#L72).

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordHelper.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordHelper.java:39]: Makes sense I was also thinking about this

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordHelper.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/JavaRecordHelper.java:56]: I would prefer to avoid the unnecessary copy in case of null, the null check is faster

File: flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/Java17RecordHelperTest.java
Comment (gyfora) [reply flink-core/src/test/java/org/apache/flink/api/java/typeutils/runtime/Java17RecordHelperTest.java:61]: I renamed this to `setField` and added javadoc, that feels more natural in the context where this is used. Given an internal class I prefer a short name + javadocs

**Comment by XComp** [comment]
After a brief discussion with @gyfora , I created FLINK-33342 to cover the missing target-version for Java17

**Comment by gyfora** [comment]
@flinkbot run azure

**Comment by gyfora** [comment]
@XComp we have a green CI run with the Java tests enabled , I will remove that commit before merging.

Do you have any further comments? I think I addressed most things

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:522]: Correct me here, if I'm wrong but I have the impression that `deserialize(DataInputView source)` can be implemented in the following way:
```java
public T deserialize(DataInputView source) throws IOException {
    deserialize(null, source);
}
```
It's a bit tricky to investigate because if conditions are differently implemented. But the code paths match with `reuse == null` as far as I can see. This would reduce the amount of code redundancy in this class tremendously (it might deserve its own commit, I guess? :thinking: ).

The same seems to be true for the `copy` command. Refactoring that would reduce the diff by quite a bit as well. WDYT?

Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:522]: tbh, I feel like the `TypeSerializer` interface could be cleaned up in this regard. Some `reuse` should be marked as `@Nullable` and each implementation should be able to handle that specific case (i.e. instantiating a new instance) rather than having two separate implementations. There would be a default implementation of `copy(T)` and `deserialize(T)` in that case. But that's out-of-scope for this PR.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:532]: `reuse` will never be `null` here. We should reiterate over the code here. We create quite a bit of redundant code. The for loop and the exception handling are the same. The only difference is how the field values are set. We could use some consumer to out-source that and move the rest of the code into its own method. WDYT?

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks, @gyfora. I did another (partial) pass over the PR. I stopped after realizing that there is quite some room for improving the `PojoSerializer` code (this is not due to your changes but the class' current state in general). It comes with duplicated code as far as I can see. The change of this PR just would make it worse). 

I documented my finding in the `deserialize` methods. The same applies to the `copy` methods. I'm curious what you think.

Additionally, there are a few review comments of mine that you didn't address but also didn't resolve. Could you either mark them as resolved w/o fixing it (which I'm also ok with if you think it's too much) or address the comment (e.g. [this one](https://github.com/apache/flink/pull/23490/files#r1365477495))? I just want to make sure that comments are not accidentally missed.

**Comment by gyfora** [comment]
> Thanks, @gyfora. I did another (partial) pass over the PR. I stopped after realizing that there is quite some room for improving the `PojoSerializer` code (this is not due to your changes but the class' current state in general. It comes with duplicated code as far as I can see. The change of this PR just would make it worse).
> 
> I documented my finding in the `deserialize` methods. The same applies to the `copy` methods. I'm curious what you think.
> 
> Additionally, there are a few review comments of mine that you didn't address but also didn't resolve. Could you either mark them as resolved w/o fixing it (which I'm also ok with if you think it's too much) or address the comment (e.g. [this one](https://github.com/apache/flink/pull/23490/files#r1365477495))? I just want to make sure that comments are not accidentally missed.

Thanks @XComp , I will try to accommodate some of your suggestions. The pojoserializer has a lot of duplication that’s true but I intentionally did not want to refactor unrelated code as this is quite performance (and otherwise) sensitive area :)

I will see what we can do with confidence but I am not going to dive too deep in the refactoring

**Comment by XComp** [comment]
> I will see what we can do with confidence but I am not going to dive too deep in the refactoring

Fair enough. It might be worth having a separate Jira issue to cover the refactoring. That would allow us to separate the Java record testing from the refactoring within the performance tests. I, probably, got too excited about reducing the code in this class.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:98]: I updated the PojoTypeinfo javadocs

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:522]: I will improve the deserialise but the copy is a bit trickier.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:448]: not relevant after refactoring

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java]: Your suspicion/caution was correct - it's actually not that straight-forward setting `reuse = null` due to the recursive call of `deserialize(DataInputView)`/`deserialize(T, DataInputView)` on the fields. -.-

Anyway, I went through the code once more to figure out how we could reduce the redundant code here a bit more:
```java
    private void deserializeFields(
            DataInputView source,
            BiConsumerWithException<Integer, Object, IllegalAccessException> fieldInitializer)
            throws IOException {
        deserializeFields(
                source,
                fieldInitializer,
                fieldIdx -> fieldSerializers[fieldIdx].deserialize(source));
    }

    private void deserializeFields(
            DataInputView source,
            T reuse,
            BiConsumerWithException<Integer, Object, IllegalAccessException> fieldInitializer)
            throws IOException {
        deserializeFields(
                source,
                fieldInitializer,
                fieldIdx -> {
                    final Object reuseFieldValue =
                            reuse == null ? null : fields[fieldIdx].get(reuse);
                    if (reuseFieldValue != null) {
                        return fieldSerializers[fieldIdx].deserialize(reuseFieldValue, source);
                    } else {
                        return fieldSerializers[fieldIdx].deserialize(source);
                    }
                });
    }

    private void deserializeFields(
            DataInputView source,
            BiConsumerWithException<Integer, Object, IllegalAccessException> fieldInitializer,
            FunctionWithException<Integer, Object, ? extends Exception> valueSupplier)
            throws IOException {
        try {
            for (int i = 0; i < numFields; i++) {
                boolean isNull = source.readBoolean();

                if (fields[i] != null && isNull) {
                    fieldInitializer.accept(i, null);
                } else if (fields[i] != null && !isNull) {
                    fieldInitializer.accept(i, valueSupplier.apply(i));
                } else if (fields[i] == null && !isNull) {
                    // read and dump a pre-existing field value
                    fieldSerializers[i].deserialize(source);
                }
            }
        } catch (IllegalAccessException e) {
            throw new RuntimeException(
                    "Error during POJO copy, this should not happen since we check the fields before.",
                    e);
        } catch (Throwable e) {
            ExceptionUtils.rethrowIOException(e);
        }
    }

```

...just as a proposal. I just saw your comment that your not too excited about using consumers. I see your point there and I'm not pushing for it. I wanted to share my findings anyway.

Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java]: oh yea you are right, I will actually revery my deserialise refactor, to avoid this unintended change. 

I will not do this refactoring now. The consumers in theory can solve this but I suspect there will be an intended performance hit. I also don't want to spend investigating that too much. Will open a follow up ticket.

**Comment by gyfora** [comment]
Updated the PR @XComp , I did not address the test naming question because I can't reply to it, I think it's fine to distinguish by prefix.

**Review by XComp — CHANGES_REQUESTED** [review]
Thanks, @gyfora for addressing my comments. There are a few issues in the code still as far as I understand. PTAL

fyi: I created FLINK-33477 for the refactoring part.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:534]: let me just revert this change, reuse may or may not be null. If you pass a non-null reuse it's not null, but otherwise createInstance returns null for records

**Comment by gyfora** [comment]
Addressed these comments @XComp

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:549]: test

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java]: > If you pass a non-null reuse it's not null, but otherwise createInstance returns null for records

true, I overlooked that `createInstance` actually returns `null` in certain situations.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java]: This is not resolved in the code. I still struggle to understand why this is not necessary to be added here :thinking:

**Review by XComp — COMMENTED** [review]
Thanks for addressing my comments. One issue is still open, though. Could you clarify why the try/catch block is not needed? I feel like I'm missing something here.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java]: I only add the try catch block where the exception is declared as thrown (where it’s necessary) it’s not thrown in this code hence no try catch block . I would remove it if I could :)

Comment (gyfora) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java]: You can try removing it but I think you will find that it’s needed

**Comment by gyfora** [comment]
@XComp i would like to merge this unless you have any further functional comments. 

Since you are working on refactoring this, you can try removing some try catch block if you feel that way, but I could not do it easily.

File: flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java
Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java]: Ok, now I see where I was wrong: I was too distracted by the redundant code that I didn't see that the `IllegalAccessException` is thrown by the `Field#set` method which is not present in the first code block. Sorry for bringing this up over and over again :facepalm:

**Review by XComp — APPROVED** [review]
Nothing to add from my side. :+1:

**Comment by grzegorz8** [comment]
@gyfora Hey! Are these benchmarks available somewhere? I was not able to find them on my own.

```
RecordSerializationBenchmark.copyPojo         thrpt   30  6825.981 ± 32.221  ops/ms
RecordSerializationBenchmark.copyRecord       thrpt   30  5949.899 ± 94.196  ops/ms
RecordSerializationBenchmark.copyPojoReuse    thrpt   30  6757.649 ± 91.903  ops/ms
RecordSerializationBenchmark.copyRecordReuse  thrpt   30  5974.339 ± 51.605  ops/ms
RecordSerializationBenchmark.copyPojoView     thrpt   30   929.945 ±  3.241  ops/ms
RecordSerializationBenchmark.copyRecordView   thrpt   30   916.406 ± 53.823  ops/ms
RecordSerializationBenchmark.readPojo         thrpt   30  1070.375 ±  5.260  ops/ms
RecordSerializationBenchmark.readRecord       thrpt   30  1041.910 ± 11.632  ops/ms
RecordSerializationBenchmark.writePojo        thrpt   30  1225.184 ±  3.824  ops/ms
RecordSerializationBenchmark.writeRecord      thrpt   30  1234.543 ±  4.738  ops/ms
```

I did my own simple benchmarks and it seems that for objects with higher number of fields, deserialization of Record is noticeably faster than deserialization of Pojo. Pojo deserialization, in turn is faster or equally fast only for a small objects.

**Pull request merged**
